helix-vcs/src/git.rs

```diff
@@ -0,0 +1,113 @@
+use std::path::Path;
+
+use git::objs::tree::EntryMode;
+use git::sec::trust::DefaultForLevel;
+use git::{Commit, ObjectId, Repository, ThreadSafeRepository};
+use git_repository as git;
+
+use crate::DiffProvider;
+
+#[cfg(test)]
+mod test;
+
+pub struct Git;
+
+impl Git {
+    fn open_repo(path: &Path, ceiling_dir: Option<&Path>) -> Option<ThreadSafeRepository> {
+        // custom open options
+        let mut git_open_opts_map = git::sec::trust::Mapping::<git::open::Options>::default();
+
+        // don't use the global git configs (not needed)
+        let config = git::permissions::Config {
+            system: false,
+            git: false,
+            user: false,
+            env: true,
+            includes: true,
+        };
+        // change options for config permissions without touching anything else
+        git_open_opts_map.reduced = git_open_opts_map.reduced.permissions(git::Permissions {
+            config,
+            ..git::Permissions::default_for_level(git::sec::Trust::Reduced)
+        });
+        git_open_opts_map.full = git_open_opts_map.full.permissions(git::Permissions {
+            config,
+            ..git::Permissions::default_for_level(git::sec::Trust::Full)
+        });
+
+        let mut open_options = git::discover::upwards::Options::default();
+        if let Some(ceiling_dir) = ceiling_dir {
+            open_options.ceiling_dirs = vec![ceiling_dir.to_owned()];
+        }
+
+        ThreadSafeRepository::discover_with_environment_overrides_opts(
+            path,
+            open_options,
+            git_open_opts_map,
+        )
+        .ok()
```

Comment:
Since the `parallel` feature isn't set in the `git-features` crate, the `ThreadSafeRepository` isn't actually thread-safe as it now uses `Rc` internally instead of `Arc`. Thus here it seems better to call `to_thread_local()` right away.

Note that should you ever plan to store/cache an opened repository, the `ThreadLocalRepository` would be it the smallest in terms of memory footprint, even though I'd not advise to cache these as this makes it easy to run into system resource limits (like open file handles).

Reply:
I think it doesn't really make a difference here because the `open_repository` function is only called once and I immediately call `to_thread_local()` on the result. I would expect that to always get inclined. 

The reason I choose to return a `ThreadSafeRepository` here is that my local implementation of caching stores a `ThreadSafeRepository` in an `Arc` and the turns that into a repository when actual operations are performed. 

Do I understand the second part of your comment correctly that caching the `ThreadSafeRepository` is the correct approach or should the thread local `Repository` be cached instead?

Reply:
> Do I understand the second part of your comment correctly that caching the ThreadSafeRepository is the correct approach or should the thread local Repository be cached instead?

Yes, that's correct. The `ThreadSafeRepository` type is solely for being passed around and if `git-features/parallel` is active, it's `Send + Sync`. This might be something to improve as doing so will also pull in dependencies to crossbeam for the scoped-thread, which is nothing you really need. On the other hand, just waiting for `std::thread::scope` to move into the MSRV should be able to shed that dependency, maybe it's alright.

Reply:
I didn't want to enable the parallel feature for this PR yet because it pulls in rayon and jwalk as extra dependencies and rayon can be a bit heavy on compile times. Because a single threaded repository is just fine for this version. I was hoping to send a PR against gitoxide to make those dependencies optional before opening the caching PR